src/hypothesis/extra/django/models.py

```diff
@@ -17,159 +17,332 @@
+                .format(
+                    model=model.__name__,
+                    field=field.name,
+                )
+            )
+        field_strategies[field.name] = strategy
     # Remove default_values so we don't try to generate anything for those.
-    result = {k: v for k, v in result.items() if v is not default_value}
-    return ModelStrategy(model, result)
+    field_strategies = {
+        field_name: strategy
+        for field_name, strategy
+        in field_strategies.items()
+        if strategy is not default_value
+    }
+    # Create the model data.
+    model_data_strategy = st.fixed_dictionaries(field_strategies)
+    return ModelStrategy(model, _db, model_data_strategy)
 
 
 class ModelStrategy(SearchStrategy):
 
-    def __init__(self, model, mappings):
+    def __init__(self, model, db, model_data_strategy):
         super(ModelStrategy, self).__init__()
         self.model = model
-        self.arg_strategy = st.fixed_dictionaries(mappings)
+        self.db = db
+        self.model_data_strategy = model_data_strategy
 
     def __repr__(self):
         return u'ModelStrategy(%s)' % (self.model.__name__,)
 
     def do_draw(self, data):
+        model_data = data.draw(self.model_data_strategy)
         try:
-            result, _ = self.model.objects.get_or_create(
-                **self.arg_strategy.do_draw(data)
-            )
-            return result
-        except IntegrityError:
+            # We need to wrap the model create in an atomic block, so
+            # we need to use the correct write database for the model.
+            db = self.db or router.db_for_write(self.model)
+            # If the save gives an IntegrityError, this will roll the
+            # savepoint back.
+            with transaction.atomic(using=db):
+                # Create the model inside the transaction, just in case it
+                # performs database actions in __init__.
+                obj = self.model(**model_data)
+                # This should catch unique checks, plus any other custom
+                # validation on the model.
+                obj.full_clean()
+                obj.save(using=db)
+            return obj
+        except DataError:
+            # Something weird happened. For example,
+            # some Postgres database encodings will refuse text that
+            # is longer than a VARCHAR *once encoded by the database*.
             data.mark_invalid()
+        except (ValidationError, IntegrityError):
+            # This might mean that a unique key way violoated.
+            # As a fallback, try to get the identical model. This is needed
+            # for example() calls, which expect to be able to call the
+            # strategey idempotently.
+            try:
+                return self.model._default_manager.get(**model_data)
+            except self.model.DoesNotExist:
+                # Something other than a unique violoation
```

Comment:
Minor: "way violoated"? Did you mean "was violated"?

In the same comment: "strategey" ~> "strategy" and "violoation" ~> "violation".